Enable central package management implicitly when Directory.Packages.props exists

jeffkl opened the pull request:
<!-- DO NOT MODIFY OR DELETE THIS TEMPLATE. IT IS USED IN AUTOMATION. -->
## Bug

<!-- Search https://github.com/NuGet/Home/issues, and create one if you can't find a suitable issue. -->
<!-- Paste the full link, like https://github.com/nuget/home/issues/1000. GitHub will render is neatly. -->
Fixes: https://github.com/NuGet/Home/issues/11834

Regression? Last working version:

## Description
<!-- Add details about the fix. Include any information that would help the maintainer review this change effective. -->
This change makes it so the presence of a `Directory.Packages.props` means that central package management is implicitly on.  I've had countless people ask over the last year why you have to have the file and explicitly set `ManagePackageVersionsCentrally` to `true`.  The logic now assumes if `Directory.Packages.props` exists then `ManagePackageVersionsCentrally` should be `true` unless the user has already set that property to a value, including if its set in `Directory.Packages.props`.  This will make it easier for customers to onboard.  


## PR Checklist

- [x] PR has a meaningful title
- [x] PR has a linked issue.
- [x] Described changes

- **Tests**
  - [x] Automated tests added
  - **OR**
  <!-- Describe why you haven't added automation. -->
  - [ ] Test exception
  - **OR**
  - [ ] N/A <!-- Infrastructure, documentation etc. -->

- **Documentation**
  <!-- Please link the PR/issue if appropriate -->
  - [ ] Documentation PR or issue filled
  - **OR**
  - [x] N/A

jeffkl committed:
Fix NuGet.exe to have NuGet.props

jeffkl committed:
Implementation

jeffkl committed:
Update existing tests

jeffkl committed:
Add test

jeffkl commented:
Team Review:  Add "breaking change" to the issue and document that setting the property explicitly is no longer needed in Version X

kartheekp-ms commented on test/NuGet.Core.FuncTests/Dotnet.Integration.Test/DotnetAddPackageTests.cs:
```diff
@@ -945,12 +924,7 @@ public async Task AddPkg_WithCPM_WithPackageReference_NoVersionOverride_NoPackag
                     packageX200);
 
-            var propsFile = @$"<Project>
-                                <PropertyGroup>
-                                    <ManagePackageVersionsCentrally>true</ManagePackageVersionsCentrally>
-                                </PropertyGroup>
```
Do we need one test to ensure CPM restore works as expected for projects where `ManagePackageVersionsCentrally` property has been configured explicitly?

donnie-msft replied:
Sounds like a good idea to me.

jeffkl replied:
This new test is doing this, `true` is the user explicitly opted in, `false` is they opted out, and `null` is they did not specify so they get the default of enabled since Directory.Packages.props exists.

https://github.com/NuGet/NuGet.Client/pull/5456/files#diff-1faf54421a99e7ac571eb0ca8dd73601dc5d38ee395668da1373fef077488154R10111-R10114

donnie-msft commented on test/NuGet.Core.FuncTests/Dotnet.Integration.Test/PackCommandTests.cs:
```diff
@@ -5246,15 +5241,12 @@ public void PackCommand_PackProjectWithCentralTransitiveDependencies(string Cent
 
                 // The test depends on the presence of these packages and their versions.
                 // Change to Directory.Packages.props when new cli that supports NuGet.props will be downloaded
```
No longer need this "todo" ?
```suggestion
```

jeffkl committed:
Address feedback

jeffkl committed:
Fix end-to-end tests

donnie-msft commented on test/NuGet.Clients.Tests/NuGet.CommandLine.Test/RestoreNETCoreTest.cs:
```diff
@@ -10118,6 +10107,80 @@ SimpleTestPackageContext createTestPackage(string name, string version, List<Sim
+                var libraries = assetsFile.Libraries.Select(l => $"{l.Name}.{l.Version}").OrderBy(n => n).ToList();
+                Assert.Equal(expectedLibraries, libraries);
+
+                var centralfileDependencyGroups = assetsFile
```
```suggestion
                var centralFileDependencyGroups = assetsFile
```

jeffkl replied:
Fixed

donnie-msft commented on test/NuGet.Clients.Tests/NuGet.CommandLine.Test/RestoreNETCoreTest.cs:
```diff
@@ -10118,6 +10107,80 @@ SimpleTestPackageContext createTestPackage(string name, string version, List<Sim
+        public async Task RestoreNetCore_CPVMProject_ManagePackageVersionsCentrally_CanBeDisabled(bool? managePackageVersionsCentrally)
+        {
+            // Arrange
+            using (var pathContext = new SimpleTestPathContext())
```
I'd prefer an "inline using" to avoid lots of indenting.

jeffkl replied:
There's around 100 of these in this test file.  I think I'll leave it as-is for now so its consistent with the rest of the test class and if we want to improve this later as a tech debt issue we can do it then.

jeffkl replied:
I have re-written this test from scratch, please re-review it.

donnie-msft commented on test/NuGet.Clients.Tests/NuGet.CommandLine.Test/RestoreNETCoreTest.cs:
```diff
@@ -10118,6 +10107,80 @@ SimpleTestPackageContext createTestPackage(string name, string version, List<Sim
+                   framework);
+
+                // the package references defined in the project should not have version unless CPM is disabled
+                var packageBNoVersion = createTestPackage("B", managePackageVersionsCentrally is null || managePackageVersionsCentrally == true ? null : "1.0.0", packagesForProject);
```
I don't see where this one is used?
```suggestion

```

jeffkl replied:
Removed

donnie-msft commented on test/NuGet.Clients.Tests/NuGet.CommandLine.Test/RestoreNETCoreTest.cs:
```diff
@@ -10118,6 +10107,80 @@ SimpleTestPackageContext createTestPackage(string name, string version, List<Sim
+                solution.Create(pathContext.SolutionRoot);
+
+                // Act
+                var r = Util.RestoreSolution(pathContext);
```
```suggestion
                var result = Util.RestoreSolution(pathContext);
```

donnie-msft commented on test/NuGet.Clients.Tests/NuGet.CommandLine.Test/RestoreNETCoreTest.cs:
```diff
@@ -10118,6 +10107,80 @@ SimpleTestPackageContext createTestPackage(string name, string version, List<Sim
+
+                // the package references defined in the project should not have version unless CPM is disabled
+                var packageBNoVersion = createTestPackage("B", managePackageVersionsCentrally is null || managePackageVersionsCentrally == true ? null : "1.0.0", packagesForProject);
+                var packageB100 = createTestPackage("B", "1.0.0", packagesForSource);
```
It would be less for me to keep track of if this variable was either an explicit type and/or the variable name indicates it's a context.

```suggestion
                var packageContextB100 = createTestPackage("B", "1.0.0", packagesForSource);
```

jeffkl replied:
This is all just copy/paste so I didn't bother cleaning up poor patterns

donnie-msft commented on test/NuGet.Clients.Tests/NuGet.CommandLine.Test/RestoreNETCoreTest.cs:
```diff
@@ -10118,6 +10107,80 @@ SimpleTestPackageContext createTestPackage(string name, string version, List<Sim
+
+                var centralfileDependencyGroups = assetsFile
+                    .CentralTransitiveDependencyGroups
+                    .SelectMany(g => g.TransitiveDependencies.Select(t => $"{g.FrameworkName}_{t.LibraryRange.Name}.{t.LibraryRange.VersionRange.OriginalString}")).ToList();
```
Can this be simplified to just `group => group.TransitiveDependencies`, then check that count without building the strings? 
Or even further, can `.CentralTransitiveDependencyGroups.Count` be asserted?

jeffkl replied:
I'm not sure, I'm going to leave it as-is for now for sake of time but this whole test class could probably use some love

donnie-msft commented on test/NuGet.Clients.Tests/NuGet.CommandLine.Test/RestoreNETCoreTest.cs:
```diff
@@ -10118,6 +10107,80 @@ SimpleTestPackageContext createTestPackage(string name, string version, List<Sim
+                };
+
+                var projectA = SimpleTestProjectContext.CreateNETCore(
+                   "projectA",
```
I believe I'm only seeing 3 spaces in the indentation (need 4)
```suggestion
                    "projectA",
```

donnie-msft commented on test/NuGet.Clients.Tests/NuGet.CommandLine.Test/RestoreNETCoreTest.cs:
```diff
@@ -10118,6 +10107,80 @@ SimpleTestPackageContext createTestPackage(string name, string version, List<Sim
+                var packageBNoVersion = createTestPackage("B", managePackageVersionsCentrally is null || managePackageVersionsCentrally == true ? null : "1.0.0", packagesForProject);
+                var packageB100 = createTestPackage("B", "1.0.0", packagesForSource);
+                var packageC100 = createTestPackage("C", "1.0.0", packagesForSource);
+                var packageC200 = createTestPackage("C", "2.0.0", packagesForSource);
```
Are you needing this one in the array, `packagesForSource`? It'd be easier for me to understand if the variable is removed if it's unused.

```suggestion
                createTestPackage("C", "2.0.0", packagesForSource);
```

jeffkl replied:
This whole test is just a copy/paste so I'm not as worried about poor patterns, do you mind yourself fixing up test classes when you work on a new test?

donnie-msft replied:
>  do you mind yourself fixing up test classes when you work on a new test?

When I write **new tests**, I try to follow our guidelines. If it means something complicated like creating a bunch of helper methods, I may or may not do that. If it's something trivial like using proper variable names or conventions, I tend to do that in my new tests. I don't tend to update entire class files (which is not what I suggested, to be clear).

donnie-msft reviewed (dismissed):
Have several nits on the tests. Sorry about not finding them earlier.

jeffkl commented:
Marking this as a draft for now since it will conflict with https://github.com/NuGet/NuGet.Client/pull/5440 which I'll merge first.  I'll re-request reviews once this is rebased and ready

jeffkl committed:
Address feedback

donnie-msft commented on test/NuGet.Clients.Tests/NuGet.CommandLine.Test/RestoreNETCoreTest.cs:
```diff
@@ -9488,7 +9484,7 @@ public async Task RestoreNetCore_CPVMProject_RemovedCentralTransitiveDependency_
 
         [Fact]
-        public async Task RestoreNetCore_CPVMProject_MoveTransitiveDependnecyToCentralFile_FailsRestore()
+        public async Task RestoreNetCore_CPVMProject_MoveTransitiveDependencyToCentralFile_FailsRestore()
```
In another comment, you said you didn't want to rename a test because you'd lose test history. Is renaming it here intentional?

jeffkl replied:
Good catch, I think this was unintentional so I'll rename it back

donnie-msft commented on test/NuGet.Clients.Tests/NuGet.CommandLine.Test/RestoreNETCoreTest.cs:
```diff
@@ -9948,13 +9938,13 @@ SimpleTestPackageContext createTestPackage(string name, string version, List<Sim
                     .SelectMany(g => g.TransitiveDependencies.Select(t => $"{g.FrameworkName}_{t.LibraryRange.Name}.{t.LibraryRange.VersionRange.OriginalString}")).ToList();
 
-                var expectedCentralfileDependencyGroups = new List<string>() { $"{framework.DotNetFrameworkName}_P.[3.0.0, )", $"{framework.DotNetFrameworkName}_S.[3.0.0, )" };
+                var expectedcentralFileDependencyGroups = new List<string>() { $"{framework.DotNetFrameworkName}_P.[3.0.0, )", $"{framework.DotNetFrameworkName}_S.[3.0.0, )" };
```
Think this rename was a mistake

```suggestion
                var expectedCentralFileDependencyGroups = new List<string>() { $"{framework.DotNetFrameworkName}_P.[3.0.0, )", $"{framework.DotNetFrameworkName}_S.[3.0.0, )" };
```

jeffkl replied:
Thanks, fixed

jeffkl committed:
Rewrite test

jeffkl committed:
Fix whitespace

jeffkl commented:
I deleted two comment threads related to the test I rewrote and left [the one behind](https://github.com/NuGet/NuGet.Client/pull/5456/files#diff-1faf54421a99e7ac571eb0ca8dd73601dc5d38ee395668da1373fef077488154) that spawned the idea with my reply indicating that the test was rewritten.

ghost commented:
This PR has been automatically marked as stale because it has no activity for **7 days**. It will be closed if no further activity occurs **within another 7 days of this comment**. If it is closed, you may reopen it anytime when you're ready again, as long as you don't delete the branch.
